src/coprocessor/dag/expr/column.rs

```diff
@@ -38,8 +38,33 @@ impl Column {
     }
 
     #[inline]
-    pub fn eval_string<'a>(&self, row: &'a [Datum]) -> Result<Option<Cow<'a, [u8]>>> {
-        row[self.offset].as_string()
+    pub fn eval_string<'a>(
+        &self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        if let Datum::Null = row[self.offset] {
+            return Ok(None);
+        }
+        if types::is_hybrid_type(self.tp.get_tp() as u8) {
+            let s = row[self.offset].to_string()?.into_bytes();
+            return Ok(Some(Cow::Owned(s)));
+        }
+
+        if !ctx.cfg.pad_char_to_full_length || self.tp.get_tp() != i32::from(types::STRING) {
+            return row[self.offset].as_string();
+        }
+
+        let res = row[self.offset].as_string()?.unwrap();
+        let cur_len = str::from_utf8(res.as_ref())?.chars().count();
+        let flen = self.tp.get_flen() as usize;
+        if flen <= cur_len {
+            return Ok(Some(res));
+        }
+        let new_len = flen - cur_len + res.len();
```

Comment:
I remembered flen is the length of bytes instead of characters. there should be issues when mixing flen with a string length. could you have a check?

Reply:
I've added it to the test.